pkg/urbit/vere/epoc.c

```diff
@@ -0,0 +1,903 @@
+
+succeed:
+  return poc_u;
+}
+
+u3_epoc*
+u3_epoc_migrate(c3_path* const       src_u,
+                const c3_path* const par_u,
+                const c3_d           cur_d,
+                u3_meta* const       met_u)
+{
+  if ( !src_u || !par_u ) {
+    goto fail;
+  }
+
+  u3_epoc* poc_u = c3_calloc(sizeof(*poc_u));
+  poc_u->fir_d   = epo_min_d;
+  poc_u->las_d   = epo_min_d - 1;
+  // Protect against unlikely case where `epo_min_d` is erroneously changed to
+  // 0.
+  c3_assert(poc_u->las_d < poc_u->fir_d); // (1)
+  poc_u->pax_u   = _epoc_path(par_u, epo_min_d);
+  mkdir(c3_path_str(poc_u->pax_u), 0700);
+
+  { // Relocate LMDB instance to the newly created epoch.
+    if ( !_move_file(src_u, poc_u->pax_u, "data.mdb") ) {
+      goto free_epoc;
+    }
+    if ( !_move_file(src_u, poc_u->pax_u, "lock.mdb") ) {
+      goto rename_data_mdb;
+    }
+  }
+
+  if ( !(poc_u->env_u = _lmdb_init(poc_u->pax_u)) ) {
+    goto rename_lock_mdb;
+  }
+
+  if ( !_lmdb_gulf(poc_u->env_u, &poc_u->fir_d, &poc_u->las_d) ) {
+    goto rename_lock_mdb;
+  }
+
+  if ( u3_epoc_is_empty(poc_u) ) {
+    fprintf(stderr,
+            "epoc: old event log at %s is unexpectedly empty\r\n",
+            c3_path_str(poc_u->pax_u));
+    goto rename_lock_mdb;
+  }
+
+  // The incremental snapshot must be up-to-date to successfully migrate.
+  if ( cur_d != poc_u->las_d ) {
+    // TODO(peter): update version number.
+    fprintf(stderr,
+            "IMPORTANT: cannot migrate the existing event log format to the\r\n"
+            "           new epoch-based event log format because the\r\n"
+            "           incremental snapshot is not up-to-date. To resolve\r\n"
+            "           this, run your ship using version 1.8 of the urbit\r\n"
+            "           binary and exit *gracefully* using Ctrl-D. Then,\r\n"
+            "           try again using the latest version of the urbit\r\n"
+            "           binary. If you have questions or concerns, please\r\n"
+            "           file an issue against\r\n"
+            "           https://github.com/urbit/urbit and assign it to\r\n"
+            "           mcevoypeter.\r\n");
+    goto rename_lock_mdb;
+  }
+
+  MDB_txn* txn_u;
+  { // Read metadata out of LMDB instance.
+    try_lmdb(mdb_txn_begin(poc_u->env_u, NULL, MDB_RDONLY, &txn_u),
+             goto rename_lock_mdb,
+             "failed to create read-only transaction");
+
+    MDB_dbi dbi_u;
+    try_lmdb(mdb_dbi_open(txn_u, "META", 0, &dbi_u),
+             goto abort_txn,
+             "failed to open META database");
+
+#define lookup_metadata(key_string, destination)                               \
+    do {                                                                       \
+      MDB_val key_u = {                                                        \
+        .mv_data = (void*)key_string,                                          \
+        .mv_size = strlen(key_string) + 1,                                     \
```

Comment:
This is wrong, the existing metadata keys are byte arrays without the null terminator.

Reply:
Thanks for pointing this out. I've corrected the issue.